contrib/cf-diag.sh

```diff
@@ -0,0 +1,240 @@
+# Skip ahead to TESTS instead.
+
+test_number=0
+
+outcome() {
+  # Accepts from one to three args:
+  # First - outcome, should be 'ok' or 'not ok'
+  # Second - optional description
+  # Third - optional directive
+  # (see TAP specification https://testanything.org/tap-specification.html)
+
+  # (Third arg not currently used.)
+
+  test_number="$((test_number+1))"
+  printf '%s %d%s%s\n' "$1" "$test_number" ${2+" - $2"} ${3+" # $3"}
+  # See https://unix.stackexchange.com/a/68488/135943
+  # for the basis of the shell trick in use here
+}
+
+succeed() {
+  outcome ok "$@"
+}
+
+fail() {
+  outcome 'not ok' "$@"
+}
+
+bail() {
+  printf 'Bail out!%s\n' ${1+" $1"}
+  exit
+}
+
+trap 'printf "1..%d\n" "$test_number"' EXIT
+
+#################################### TESTS ####################################
+
+# I recommend custom checks be added only at the end.
+# Skip ahead to CUSTOM TESTS.
+
+# Show host key (digest)
+printf '# '
+cf-key -p /var/cfengine/ppkeys/localhost.pub 2>/dev/null || echo SHA not available
+
+# VERSION CHECKS
+# makes use of 'ok_versions' defined at top of script
+
+version_data="$(cf-agent -V 2>/dev/null)"
+if [ "$?" = 127 ]; then # Don't go further if cf-agent missing
+  bail "$(cf-agent -V 2>&1 1>/dev/null)"
+fi
+
+# Beware literal newlines; don't try adding trailing comments!
+core_version_line="${version_data%
+*}" # e.g. 'CFEngine Core 3.7.3'
+core_version="${core_version_line##* }" # e.g. '3.7.3'
+enterprise_version_line="${version_data#*
+}" # e.g. 'CFEngine Enterprise 3.7.3'
+enterprise_version="${enterprise_version_line##* }" # e.g. '3.7.3'
+
+desc="Expected Core version"
+if printf '%s\n' "$core_version" | grep -qFx $ok_versions; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+desc="Expected Enterprise version"
+if printf '%s\n' "$enterprise_version" | grep -qFx $ok_versions; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+desc="Core version matches Enterprise version"
+# In pathological cases these may not match.
+# The individual checks for Core and Enterprise versions ok
+# may not detect these pathological cases if e.g. 3.7.3 and 3.10.2
+# are both allowed ("okay") versions, so we test it separately.
+if [ "$core_version" = "$enterprise_version" ]; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+# Show actual version data
+# (Can be commented out if you only allow one version anyway.)
+printf '# %s\n' "$core_version_line" "$enterprise_version_line"
+
+# CFENGINE FUNCTIONING
+
+# Check all three daemons as separate tests
+for d in cf-execd cf-monitord cf-serverd; do
+  desc="$d running"
+  if pgrep "$d" >/dev/null 2>/dev/null; then
+    succeed "$desc"
+  else
+    fail "$desc"
+  fi
+done
+
+# Test for recent run:
+
+# 15 minutes is three times the out-of-the-box default run interval (5 min)
+# However, as noted by Neil Watson
+# (in this ticket: https://tracker.mender.io/browse/CFE-769),
+# most large CFEngine installations use longer (less frequent) run intervals.
```

Comment:
s/most/some/

I am actually quite surprised at the number of large installs that do run a 5-minute interval. They use other methodologies to do parts of the policies at different intervals.